File: src/main/java/view/PluginSettingsConfigurable.java
package view;

import com.intellij.openapi.options.Configurable;
import com.intellij.openapi.ui.ComboBox;
import com.intellij.openapi.ui.Messages;
import com.intellij.ui.JBColor;
import com.intellij.util.ui.JBUI;
import inputmethod.InputMethodSwitchStrategy;
import inputmethod.StrategyFactory;

import javax.swing.*;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.util.Objects;

public class PluginSettingsConfigurable implements Configurable {
    private static final String DEFAULT_STRATEGY = "UIAutomationSwitcher";
    // 默认颜色配置
    private static final Color DEFAULT_EN_COLOR = new JBColor(new Color(255, 0, 0, 255), new Color(255, 0, 0, 255));   // 红色
    private static final Color DEFAULT_ZH_COLOR = new JBColor(new Color(0, 150, 255, 255), new Color(0, 150, 255, 255)); // 蓝色

    private JPanel settingsPanel;
    private JComboBox<String> strategyComboBox;
    private JTextField enColorField;
    private JTextField zhColorField;

    private final SettingsState state;

    public PluginSettingsConfigurable() {
        this.state = SettingsState.getInstance();
        if (this.state == null) {
            throw new IllegalStateException("SettingsState must not be null");
        }
    }

    @Override
    public JComponent createComponent() {
        settingsPanel = new JPanel();
        settingsPanel.setLayout(new BoxLayout(settingsPanel, BoxLayout.Y_AXIS));
        // 1. 切换策略选择组件
        JPanel strategyPanel = createStrategyPanel();
        settingsPanel.add(strategyPanel);
        // 2. 光标颜色配置组件
//        settingsPanel.add(createColorConfigPanel(" 英文光标颜色:", "enColor"));
//        settingsPanel.add(createColorConfigPanel(" 中文光标颜色:", "zhColor"));

        // 3. 操作按钮
        JPanel buttonPanel = createButtonPanel();
        settingsPanel.add(buttonPanel);

        return settingsPanel;
    }

    // 创建策略选择面板
    private JPanel createStrategyPanel() {
        JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
        panel.add(new JLabel("切换策略:"));
        strategyComboBox = new ComboBox<>(new String[]{
                "UIAutomationSwitcher",
                "InputMethodChecker"
        });
        strategyComboBox.setRenderer(new DefaultListCellRenderer() {
            @Override
            public Component getListCellRendererComponent(JList<?> list, Object value, int index,
                                                          boolean isSelected, boolean cellHasFocus) {
                String display = value.toString()
                        .replace("UIAutomationSwitcher", "UIAutomation模拟点击托盘输入法按钮(默认)")
                        .replace("InputMethodChecker", "shift快捷键(兼容性好，可能误触发事件)");
                return super.getListCellRendererComponent(list, display, index, isSelected, cellHasFocus);
            }
        });
        panel.add(strategyComboBox);
        return panel;
    }

    // 创建颜色配置面板
    private JPanel createColorConfigPanel(String label, String type) {
        JPanel panel = new JPanel(new FlowLayout(FlowLayout.LEFT));
        panel.add(new JLabel(label));

        JTextField colorField = new JTextField(7);
        colorField.setEditable(false);

        JButton colorBtn = new JButton("选择");
        colorBtn.setMargin(JBUI.insets(2, 8));

        if ("enColor".equals(type)) {
            enColorField = colorField;
//            enColorBtn = colorBtn;
            initColorField(colorField, state.englishCursorColor, DEFAULT_EN_COLOR);
        } else {
            zhColorField = colorField;
//            zhColorBtn = colorBtn;
            initColorField(colorField, state.chineseCursorColor, DEFAULT_ZH_COLOR);
        }

        colorBtn.addActionListener(e -> chooseColor(colorField));
        panel.add(colorField);
        panel.add(colorBtn);

        return panel;
    }

    // 初始化颜色文本框
    private void initColorField(JTextField field, Color savedColor, Color defaultColor) {
        Color color = savedColor != null ? savedColor : defaultColor;
        field.setText(colorToHex(color));
        field.setBackground(color);
        field.setForeground(getContrastColor(color));  // 自动适配文本颜色
    }

    // 创建按钮面板
    private JPanel createButtonPanel() {
        JPanel buttonPanel = new JPanel(new FlowLayout(FlowLayout.LEFT));
        buttonPanel.setBorder(BorderFactory.createEmptyBorder(15, 0, 5, 0));

        JButton testInputSwitchBut = new JButton("测试输入法切换");
        testInputSwitchBut.addActionListener(this::inputSwitch);
        buttonPanel.add(testInputSwitchBut);

        return buttonPanel;
    }

    private void inputSwitch(ActionEvent e) {
        try {
            String strategyName = (String) strategyComboBox.getSelectedItem();
            InputMethodSwitchStrategy strategy = StrategyFactory.createStrategy(strategyName);
            strategy.change();
        } catch (Exception ex) {
            Messages.showErrorDialog(" 切换失败: " + ex.getMessage(), "操作异常");
        }
    }

    // 颜色选择对话框
    private void chooseColor(JTextField targetField) {
        Color initial = hexToColor(targetField.getText());
        Color color = JColorChooser.showDialog(
                settingsPanel,
                "选择光标颜色",
                initial != null ? initial : JBColor.WHITE
        );

        if (color != null) {
            targetField.setText(colorToHex(color));
            targetField.setBackground(color);
            targetField.setForeground(getContrastColor(color));
        }
    }

    // 颜色转换工具方法
    private String colorToHex(Color color) {
        return String.format("#%02x%02x%02x",
                color.getRed(),
                color.getGreen(),
                color.getBlue()
        );
    }

    private Color hexToColor(String hex) {
        try {
            return Color.decode(hex);
        } catch (Exception e) {
            return null;
        }
    }

    // 获取对比色（自动选择白/黑文字）
    private Color getContrastColor(Color color) {
        double luminance = (0.299 * color.getRed() +
                0.587 * color.getGreen() +
                0.114 * color.getBlue()) / 255;
        return luminance > 0.5 ? JBColor.BLACK : JBColor.WHITE;
    }

    @Override
    public boolean isModified() {
        boolean strategyModified = !Objects.equals(strategyComboBox.getSelectedItem(), state.strategyClass);

        boolean enColorModified = !enColorField.getText().equals(
                state.englishCursorColor != null ?
                        colorToHex(state.englishCursorColor) :
                        colorToHex(DEFAULT_EN_COLOR)
        );

        boolean zhColorModified = !zhColorField.getText().equals(
                state.chineseCursorColor != null ?
                        colorToHex(state.chineseCursorColor) :
                        colorToHex(DEFAULT_ZH_COLOR)
        );

        return strategyModified || enColorModified || zhColorModified;
    }

    @Override
    public void apply() {
        if (state != null) {
            // 保存策略配置
            state.strategyClass = (String) strategyComboBox.getSelectedItem();

            // 保存颜色配置
            state.englishCursorColor = hexToColor(enColorField.getText());
            state.chineseCursorColor = hexToColor(zhColorField.getText());
        } else {
            Messages.showWarningDialog("Err", "系统警告");
        }
    }

    @Override
    public void reset() {
        if (state != null) {
            // 重置策略选择
            strategyComboBox.setSelectedItem(
                    state.strategyClass != null ? state.strategyClass : DEFAULT_STRATEGY);

            // 重置颜色配置
            initColorField(enColorField, state.englishCursorColor, DEFAULT_EN_COLOR);
            initColorField(zhColorField, state.chineseCursorColor, DEFAULT_ZH_COLOR);
        }
    }

    @Override
    public String getDisplayName() {
        return "智能输入法切换插件";
    }
}

File: src/main/kotlin/view/PluginSettingsConfigurable.kt
package view

import com.intellij.ide.BrowserUtil
import com.intellij.openapi.options.Configurable
import com.intellij.openapi.ui.ComboBox
import com.intellij.openapi.ui.Messages
import com.intellij.openapi.ui.ValidationInfo
import com.intellij.openapi.ui.ComponentValidator
import com.intellij.openapi.util.Disposer
import inputmethod.switcher.SwitcherStrategyFactory
import java.awt.Component
import java.awt.Dimension
import javax.swing.*
import java.util.Objects
import com.intellij.ui.dsl.builder.panel
import com.intellij.ui.ColorPanel
import com.intellij.ui.JBColor
import com.intellij.ui.components.JBTextField
import java.awt.FlowLayout
import javax.swing.JComponent
import com.intellij.util.ui.JBUI
import editoraction.FocusHandel.FocusArea
import enums.InputState
import state.SettingsState
import java.awt.Cursor
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import java.awt.event.MouseEvent
import java.awt.event.MouseAdapter

class PluginSettingsConfigurable() : Configurable {

    private var switchLabel = "切换策略:"
    private var tipLabel = "提示策略:"

    private var inputSwitchStrategyComboBox: JComboBox<String>? = null
    private var switchingStrategyComboBox: JComboBox<String>? = null

//    private var editorComboBox: JComboBox<InputState>? = null
//    private var renameDialogComboBox: JComboBox<InputState>? = null
//    private var terminalComboBox: JComboBox<InputState>? = null
//    private var searchComboBox: JComboBox<InputState>? = null
//    private var otherComboBox: JComboBox<InputState>? = null


    private var zhCursorColorPanel: ColorPanel? = null
    private var enCursorColorPanel: ColorPanel? = null
    private val state = SettingsState.getInstance() ?: throw IllegalStateException("SettingsState must not be null")
    private var balloonDurationField: JBTextField? = null
    private var balloonDurationValidator: ComponentValidator? = null

    override fun createComponent(): JComponent {
        return panel {
            group("输入法设置") {
                row {
                    label(switchLabel)
                    cell(
                        createStrategyComboBox(
                            items = arrayOf("UIAutomationSwitcher", "KeyboardSwitcher"),
                            displayMap = mapOf(
                                "UIAutomationSwitcher" to "UIAutomation模拟点击托盘输入法按钮(默认)",
                                "KeyboardSwitcher" to "shift快捷键(兼容性好，可能误触发事件)"
                            ),
                            selected = state.inputSwitchStrategyClass!!
                        )
                    ).applyToComponent {
                        inputSwitchStrategyComboBox = this
                    }
                    cell(createTestButton("测试输入法切换") { inputSwitch() })
                }
            }

            group("提示设置") {
                row {
                    label(tipLabel)
                    cell(
                        createStrategyComboBox(
                            items = arrayOf("CursorColorStrategy", "BalloonStrategy"),
                            displayMap = mapOf(
                                "BalloonStrategy" to "气泡提示",
                                "CursorColorStrategy" to "光标颜色"
                            ),
                            selected = state.switchingStrategyClass!!
                        )
                    ).applyToComponent {
                        switchingStrategyComboBox = this
                        addActionListener { updateDynamicSettings() }
                    }
                }

                // 动态设置区域
                row {
                    cell(createDynamicSettingsPanel())
                }
            }

//            group("场景默认输入法设置") {
//                row {
//                    label(FocusArea.EDITOR.name)
//                    cell(
//                        createStrategyComboBox(
//                            items = arrayOf(InputState.CHINESE, InputState.ENGLISH),
//                            displayMap = mapOf(
//                                InputState.CHINESE to InputState.CHINESE.name,
//                                InputState.ENGLISH to InputState.ENGLISH.name
//                            ),
//                            selected = state.editorInputState!!
//                        )
//                    ).applyToComponent {
//                        editorComboBox = this
//                    }
//                }
//                row {
//                    label(FocusArea.RENAME_DIALOG.name)
//                    cell(
//                        createStrategyComboBox(
//                            items = arrayOf(InputState.CHINESE, InputState.ENGLISH),
//                            displayMap = mapOf(
//                                InputState.CHINESE to InputState.CHINESE.name,
//                                InputState.ENGLISH to InputState.ENGLISH.name
//                            ),
//                            selected = state.renameDialogInputState!!
//                        )
//                    ).applyToComponent {
//                        renameDialogComboBox = this
//                    }
//                }
//                row {
//                    label(FocusArea.TERMINAL.name)
//                    cell(
//                        createStrategyComboBox(
//                            items = arrayOf(InputState.CHINESE, InputState.ENGLISH),
//                            displayMap = mapOf(
//                                InputState.CHINESE to InputState.CHINESE.name,
//                                InputState.ENGLISH to InputState.ENGLISH.name
//                            ),
//                            selected = state.terminalInputState!!
//                        )
//                    ).applyToComponent {
//                        terminalComboBox = this
//                    }
//                }
//                row {
//                    label(FocusArea.SEARCH.name)
//                    cell(
//                        createStrategyComboBox(
//                            items = arrayOf(InputState.CHINESE, InputState.ENGLISH),
//                            displayMap = mapOf(
//                                InputState.CHINESE to InputState.CHINESE.name,
//                                InputState.ENGLISH to InputState.ENGLISH.name
//                            ),
//                            selected = state.searchInputState!!
//                        )
//                    ).applyToComponent {
//                        searchComboBox = this
//                    }
//                }
//            }
            separator(JBColor.GRAY)  // 添加分隔线
            row {
                cell(JPanel(FlowLayout(FlowLayout.RIGHT, 10, 5)).apply {
                    border = JBUI.Borders.empty(5, 0) // 添加上下内边距
                    add(JLabel("若插件为您带来便利，请给个 Star ⭐\uFE0F 支持！您的认可是我们持续优化的动力！").apply {
                        foreground = JBColor.GRAY
                    })
                })
            }
            row {
                cell(createFooterPanel())
            }
        }.apply {
            // 初始加载时更新动态设置
            updateDynamicSettings()
        }
    }

    private fun createFooterPanel(): JPanel {
        return JPanel(FlowLayout(FlowLayout.RIGHT, 10, 5)).apply {
            border = JBUI.Borders.empty(5, 0)

            val label = JLabel("使用说明")
            label.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
            label.foreground = JBColor(0x0066CC, 0x3399FF)
            label.toolTipText = "使用插件前系统需要配置的步骤"

            label.addMouseListener(object : MouseAdapter() {
                override fun mouseClicked(e: MouseEvent) {
                    BrowserUtil.browse("https://github.com/zy1263188600/key-switch?tab=readme-ov-file#%EF%B8%8F-%E6%B3%A8%E6%84%8F%E4%BA%8B%E9%A1%B9")
                }

                override fun mouseEntered(e: MouseEvent) {
                    label.foreground = JBColor.BLUE
                    label.text = "<html><u>使用说明</u></html>"
                }

                override fun mouseExited(e: MouseEvent) {
                    label.foreground = JBColor(0x0066CC, 0x3399FF)
                    label.text = "使用说明"
                }
            })
            add(label)

            val starLabel = JLabel("Star支持")
            starLabel.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
            starLabel.foreground = JBColor(0x0066CC, 0x3399FF)
            starLabel.toolTipText = "前往GitHub仓库支持我们"

            starLabel.addMouseListener(object : MouseAdapter() {
                override fun mouseClicked(e: MouseEvent) {
                    BrowserUtil.browse("https://github.com/zy1263188600/key-switch")
                }

                override fun mouseEntered(e: MouseEvent) {
                    starLabel.foreground = JBColor.BLUE
                    starLabel.text = "<html><u>Star支持</u></html>"
                }

                override fun mouseExited(e: MouseEvent) {
                    starLabel.foreground = JBColor(0x0066CC, 0x3399FF)
                    starLabel.text = "Star支持"
                }
            })
            add(starLabel)

            val label2 = JLabel("五星好评")
            label2.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
            label2.foreground = JBColor(0x0066CC, 0x3399FF)
            label2.toolTipText = "前往jetbrains商店支持我们"

            label2.addMouseListener(object : MouseAdapter() {
                override fun mouseClicked(e: MouseEvent) {
                    BrowserUtil.browse("https://plugins.jetbrains.com/plugin/28418-key-switch")
                }

                override fun mouseEntered(e: MouseEvent) {
                    label2.foreground = JBColor.BLUE
                    label2.text = "<html><u>五星好评</u></html>"
                }

                override fun mouseExited(e: MouseEvent) {
                    label2.foreground = JBColor(0x0066CC, 0x3399FF)
                    label2.text = "五星好评"
                }
            })
            add(label2)


        }
    }

    private fun createDynamicSettingsPanel(): JPanel {
        return JPanel().apply {
            layout = BoxLayout(this, BoxLayout.Y_AXIS)

            // 光标颜色设置 - 使用颜色选择面板
            add(JPanel(FlowLayout(FlowLayout.LEFT)).apply {
                add(JLabel("中文光标颜色:"))
                add(ColorPanel().apply {
                    zhCursorColorPanel = this
                    preferredSize = Dimension(80, preferredSize.height)
                    selectedColor = state.zhCursorColor ?: JBColor.RED
                })
                add(JLabel("英文光标颜色:"))
                add(ColorPanel().apply {
                    enCursorColorPanel = this
                    preferredSize = Dimension(80, preferredSize.height)
                    selectedColor = state.enCursorColor ?: JBColor.BLUE
                })
            })

            add(JPanel(FlowLayout(FlowLayout.LEFT)).apply {
                border = JBUI.Borders.emptyTop(5) // 添加顶部间距
                add(JLabel("气泡持续时间(ms):"))
                add(JBTextField().apply {
                    balloonDurationField = this
                    text = state.balloonDuration?.toString() ?: "500"
                    columns = 6

                    addKeyListener(object : KeyAdapter() {
                        override fun keyReleased(e: KeyEvent?) {
                            if (switchingStrategyComboBox?.selectedItem == "BalloonStrategy") {
                                balloonDurationValidator?.revalidate()
                            }
                        }
                    })

                    fun validate(): ValidationInfo? {
                        if (switchingStrategyComboBox?.selectedItem == "BalloonStrategy") {
                            val error = validateBalloonDuration()
                            if (error != null) {
                                return ValidationInfo(error, balloonDurationField)
                            }
                        }
                        return null
                    }

                    val validatorsDisposable = Disposer.newDisposable()
                    // 创建验证器
                    balloonDurationValidator =
                        ComponentValidator(validatorsDisposable).withValidator(::validate).installOn(this)
                })
            })
        }
    }

    private fun validateBalloonDuration(): String? {
        val text = balloonDurationField?.text ?: ""
        if (text.isBlank()) {
            return "请输入持续时间"
        }
        val value = try {
            text.toInt()
        } catch (_: NumberFormatException) {
            return "必须输入数字"
        }
        if (value !in 1..10000) {
            return "请输入1-10000之间的数字"
        }
        return null
    }

    private fun updateDynamicSettings() {
        val strategy = switchingStrategyComboBox?.selectedItem?.toString() ?: return

        // 控制组件可见性
        zhCursorColorPanel?.parent?.isVisible = strategy == "CursorColorStrategy"
        enCursorColorPanel?.parent?.isVisible = strategy == "CursorColorStrategy"

        // 控制气泡持续时间组件的可见性
        balloonDurationField?.parent?.isVisible = strategy == "BalloonStrategy"

        // 策略切换时重新验证
        if (strategy == "BalloonStrategy") {
            balloonDurationValidator?.revalidate()
        } else {
            balloonDurationValidator?.updateInfo(null) // 清除错误状态
        }

        // 设置初始值
        zhCursorColorPanel?.selectedColor = state.zhCursorColor ?: JBColor.RED
        enCursorColorPanel?.selectedColor = state.enCursorColor ?: JBColor.BLUE
        balloonDurationField?.text = state.balloonDuration?.toString() ?: "500"
    }

    private fun createTestButton(text: String, action: () -> Unit): JButton {
        return JButton(text).apply {
            preferredSize = Dimension(120, preferredSize.height)
            addActionListener { action() }
        }
    }

    private inline fun <reified T> createStrategyComboBox(
        items: Array<T>,
        displayMap: Map<T, String> = emptyMap(),
        selected: T
    ): ComboBox<T> {
        return ComboBox(items).apply {
            renderer = object : DefaultListCellRenderer() {
                override fun getListCellRendererComponent(
                    list: JList<*>?, value: Any?, index: Int,
                    isSelected: Boolean, cellHasFocus: Boolean
                ): Component {
                    // 安全处理类型转换
                    val item = value as? T
                    val displayText = item?.let { displayMap[it] } ?: value?.toString()
                    return super.getListCellRendererComponent(
                        list, displayText, index, isSelected, cellHasFocus
                    )
                }
            }
            selectedItem = selected
        }
    }

    private fun inputSwitch() {
        try {
            val strategyName = inputSwitchStrategyComboBox!!.selectedItem as String
            val strategy = SwitcherStrategyFactory.createStrategy(strategyName)
            strategy.change()
        } catch (ex: Exception) {
            Messages.showErrorDialog("   切换失败: ${ex.message}", "操作异常")
        }
    }

    override fun isModified(): Boolean {
        if (balloonDurationField?.isVisible == true) {
            validateBalloonDuration() == null
        }
        return !Objects.equals(inputSwitchStrategyComboBox!!.selectedItem, state.inputSwitchStrategyClass) ||
                !Objects.equals(switchingStrategyComboBox!!.selectedItem, state.switchingStrategyClass) ||

//                !Objects.equals(editorComboBox!!.selectedItem, state.editorInputState) ||
//                !Objects.equals(renameDialogComboBox!!.selectedItem, state.renameDialogInputState) ||
//                !Objects.equals(terminalComboBox!!.selectedItem, state.terminalInputState) ||
//                !Objects.equals(searchComboBox!!.selectedItem, state.searchInputState) ||

                !Objects.equals(zhCursorColorPanel?.selectedColor, state.zhCursorColor) ||
                !Objects.equals(enCursorColorPanel?.selectedColor, state.enCursorColor) ||
                (balloonDurationField?.isVisible == true && balloonDurationField?.text != state.balloonDuration?.toString())


    }

    override fun apply() {
        state.inputSwitchStrategyClass = inputSwitchStrategyComboBox!!.selectedItem as String
        state.switchingStrategyClass = switchingStrategyComboBox!!.selectedItem as String

//        state.editorInputState = editorComboBox!!.selectedItem as InputState
//        state.renameDialogInputState = renameDialogComboBox!!.selectedItem as InputState
//        state.terminalInputState = terminalComboBox!!.selectedItem as InputState
//        state.searchInputState = searchComboBox!!.selectedItem as InputState

        state.zhCursorColor = zhCursorColorPanel?.selectedColor.let { it?.let { regular -> JBColor(regular, it) } }
        state.enCursorColor = enCursorColorPanel?.selectedColor.let { it?.let { regular -> JBColor(regular, it) } }

        if (switchingStrategyComboBox?.selectedItem == "BalloonStrategy") {
            state.balloonDuration = try {
                balloonDurationField?.text?.toInt() ?: 2000
            } catch (e: NumberFormatException) {
                2000
            }
        }
    }

    override fun reset() {
        inputSwitchStrategyComboBox!!.selectedItem = state.inputSwitchStrategyClass!!
        switchingStrategyComboBox!!.selectedItem = state.switchingStrategyClass!!

//        editorComboBox!!.selectedItem = state.editorInputState!!
//        renameDialogComboBox!!.selectedItem = state.renameDialogInputState!!
//        terminalComboBox!!.selectedItem = state.terminalInputState!!
//        searchComboBox!!.selectedItem = state.searchInputState!!

        zhCursorColorPanel?.selectedColor = state.zhCursorColor ?: JBColor.RED
        enCursorColorPanel?.selectedColor = state.enCursorColor ?: JBColor.BLUE
        balloonDurationField?.text = state.balloonDuration?.toString() ?: "2000"

        updateDynamicSettings()
    }

    override fun getDisplayName() = "智能输入法切换插件"
}